File: Low Level Design (LLD)/DesignPatterns/3. Structural/Iterator/Cpp/src/ConcreteBookIterator.cpp
#include "../include/ConcreteBookIterator.h"

ConcreteBookIterator::ConcreteBookIterator(const std::vector<Book*>& books)
    : books(books), index(0) {}

bool ConcreteBookIterator::hasNext() {
    return index < books.size();
}

Book* ConcreteBookIterator::next() {
    return books[index++];
}

File: Low Level Design (LLD)/DesignPatterns/3. Structural/Iterator/Java/ConcreteBookIterator.java
import java.util.List;

public class ConcreteBookIterator implements BookIterator {
    private List<Book> books;
    private int position = 0;

    public ConcreteBookIterator(List<Book> books) {
        this.books = books;
    }

    @Override
    public boolean hasNext() {
        return position < books.size();
    }

    @Override
    public Book next() {
        return books.get(position++);
    }
}
